src/server.c

```diff
@@ -572,27 +572,33 @@ static void monitor_cb(uv_prepare_t *monitor)
 {
 	struct dqlite_node *d = monitor->data;
 	int state = raft_state(&d->raft);
-	/*
 	queue *head;
 	struct conn *conn;
-	*/
+	struct gateway *g;
 
 	if (state == RAFT_UNAVAILABLE) {
 		return;
 	}
 
-	/* TODO: we should shutdown clients that are performing SQL requests,
-	 * but not the ones which are doing management-requests, such as
-	 * transfer leadership.  */
-	/*
 	if (d->raft_state == RAFT_LEADER && state != RAFT_LEADER) {
+		tracef("leadership lost, close leader connections");
 		QUEUE__FOREACH(head, &d->conns)
 		{
 			conn = QUEUE__DATA(head, struct conn, queue);
-			conn__stop(conn);
+			g = &conn->gateway;
+			if (g == NULL || g->leader == NULL) {
+				continue;
+			}
+
+			if (g->req != NULL && g->stmt == NULL) {
+				tracef("delayed close of ongoing non-sql request");
+				g->leadership_lost = true;
+			} else {
+				tracef("immediate close of ongoing sql request or idle connection g->sql:%p", g->sql);
+				conn__stop(conn);
+			}
```

Comment:
Another little nit: perhaps you could encapsulate the logic inside this loop into a public ```conn``` method, so details about the ```gateway``` object are not exposed (if you notice, there are no references to the ```gateway``` in ```server.c```).

For example in ```conn.h``` there could be something like:

```C
/**
 * Called in case Raft leadership gets lost.
 *
 * If the connection is serving a SQL request, it will be closed immediately,
 * otherwise it will be closed as soon as the request completes.
 */
void conn__leadership_lost(struct conn *c);
```

which basically contains the logic inside the loop body.

Reply:
Ok, makes sense.